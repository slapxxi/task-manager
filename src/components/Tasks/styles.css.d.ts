export const tasks: string;
export const taskHeader: string;
export const taskTitle: string;
export const indicator: string;
export const taskDetails: string;
export const taskDescription: string;
export const taskDescriptionCompleted: string;
export const taskFooter: string;
export const checkbox: string;
export const icon: string;
export const detailsIcon: string;
export const footerIcon: string;
export const footerIconDangerous: string;
export const tags: string;
export const footerTags: string;
export const tag: string;
