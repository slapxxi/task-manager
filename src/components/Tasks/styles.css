.tasks {
  max-width: 640px;
  margin: 0 auto;
  padding: 0;
  list-style: none;
}

.taskHeader {
  display: flex;
  align-items: flex-start;
}

.taskTitle {
  flex: 1;
  margin: 0 1em;
}

.taskTitle:hover > .indicator {
  --primary-color: #778;
}

.taskDescription {
  color: #556;
}

.taskDescriptionCompleted {
  word-wrap: normal;
  margin: 0;
  padding-bottom: 4px;
  border-right: 0;
  border-left: 0;
  font-family: inherit;
  line-height: 1.4;
  white-space: pre-wrap;
}

.taskFooter {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  flex-wrap: wrap;
  padding: 30px 0 0;
}

.checkbox {
  position: relative;
  width: 22px;
  height: 22px;
  outline: none;
  border: 0;
  background: none;
  cursor: pointer;

  appearance: none;
}

.indicator {
  --primary-color: #bbc;
  margin-left: 5px;
}

.icon {
  --primary-color: #9a9aaa;
  margin-left: 10px;
  cursor: pointer;
}

.tagIcon {
  transform: rotate(-35deg);
  composes: icon;
}

.icon:hover {
  --primary-color: #16f;
}

.detailsIcon {
  width: 14px;
  padding: 5px 0 5px;
  composes: icon;
}

.footerIcon {
  margin-left: 15px;
  composes: icon;
}

.footerIconDangerous {
  composes: footerIcon;
}

.footerIconDangerous:hover {
  --primary-color: tomato;
}

.checkbox::after {
  position: absolute;
  top: 0;
  width: 22px;
  height: 22px;
  background: url('../../assets/roundcheckbox.svg') no-repeat;
  background-position: center center;
  content: '';
}

.checkbox:checked::after {
  background: url('../../assets/roundcheckbox_checked.svg') no-repeat;
}

.footerTags {
  flex: 1;
  flex-basis: 100%;
  composes: tags from '../Tags/styles.css';
}
