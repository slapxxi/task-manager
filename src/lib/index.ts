import combineClassNames from './combineClassNames';
import describeArc from './describeArc';
import fromEpochTime from './fromEpochTime';
import Keys from './keys';
import polarToCartesian from './polarToCartesian';
import createProject from './projects/createProject';
import Regex from './regex';
import createTag from './tags/createTag';
import assignToProject from './tasks/assignToProject';
import createSubtask from './tasks/createSubtask';
import createTask from './tasks/createTask';
import resetDeadline from './tasks/resetDeadline';
import setDeadline from './tasks/setDeadline';
import tagTask from './tasks/tagTask';
import toggleTask from './tasks/toggleTask';
import themeToCSSVariables from './themeToCSSVariables';
import toEpochTime from './toEpochTime';

export {
  assignToProject,
  combineClassNames,
  createProject,
  createSubtask,
  createTag,
  createTask,
  fromEpochTime,
  Keys,
  polarToCartesian,
  Regex,
  resetDeadline,
  setDeadline,
  tagTask,
  themeToCSSVariables,
  describeArc,
  toEpochTime,
  toggleTask,
};
