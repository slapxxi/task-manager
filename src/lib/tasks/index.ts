import createTask from './createTask';
import isValidTask from './isValidTask';
import toggleTask from './toggleTask';

export { createTask, toggleTask, isValidTask };
