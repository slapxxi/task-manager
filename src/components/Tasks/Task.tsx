import { setDeadline, tagTask, toggleTask } from '@lib';
import {
  Calendar,
  DateTime,
  Icon,
  SubtasksEditor,
  TagsEditor,
} from '@local/components';
import { Subtask, Tag, Task as ITask, UserCreatedTag } from '@local/types';
import { includes, isEmpty, isEqual } from 'lodash';
import * as React from 'react';
import posed from 'react-pose';
import styled from 'styled-components';
import arrow_down from '../../assets/arrow_down.svg';
import arrow_up from '../../assets/arrow_up.svg';
import flag from '../../assets/flag.svg';
import list from '../../assets/list.svg';
import tag from '../../assets/tag.svg';
import trashbin from '../../assets/trashbin.svg';
import TextArea from '../TextArea/TextArea';
import styles from './styles.css';

enum Mode {
  default,
  editing,
}

interface Props {
  task: ITask;
  expand?: boolean;
  confirmDelete?: boolean;
  onEdit?: (task: ITask) => void;
  onDelete?: (task: ITask) => void;
  onExpand?: (expand: boolean) => void;
}

interface State {
  mode: Mode;
  showCalendar: boolean;
}

class Task extends React.Component<Props, State> {
  public state = { mode: Mode.default, showCalendar: false };

  public shouldComponentUpdate(nextProps: Props, nextState: State) {
    if (nextState !== this.state) {
      return true;
    }
    if (!isEqual(nextProps, this.props)) {
      return true;
    }
    return false;
  }

  public handleToggle = () => {
    const { task, onEdit } = this.props;
    if (onEdit) {
      onEdit(toggleTask(task));
    }
  };

  public handleChangeDescription = (value: string) => {
    const { task, onEdit } = this.props;
    if (onEdit) {
      onEdit({ ...task, description: value });
    }
  };

  public handleChangeTitle = (title: string) => {
    const { task, onEdit } = this.props;
    if (onEdit) {
      onEdit({ ...task, title });
    }
  };

  public handleToggleDetails = () => {
    if (this.props.onExpand) {
      this.props.onExpand(!this.props.expand);
    }
  };

  public handleExpand = () => {
    if (this.props.onExpand) {
      this.props.onExpand(true);
    }
  };

  public handleToggleCalendar = () => {
    this.setState((state) => ({ showCalendar: !state.showCalendar }));
  };

  public handleAddTag = (newTag: UserCreatedTag) => {
    if (
      !includes(
        this.props.task.tags.map((t) => t.name.toLowerCase()),
        newTag.name,
      )
    ) {
      if (this.props.onEdit) {
        const { task } = this.props;
        this.props.onEdit(tagTask(task, newTag));
      }
    }
  };

  public handleRemoveTags = (tags: Tag[]) => {
    if (this.props.onEdit) {
      this.props.onEdit({
        ...this.props.task,
        tags: [
          ...(this.props.task.tags as Tag[]).filter(
            (propTag) => !includes(tags.map((t) => t.id), propTag.id),
          ),
        ],
      });
    }
  };

  public handleDelete = () => {
    const { onExpand, onDelete, confirmDelete } = this.props;
    if (onDelete) {
      if (confirmDelete) {
        const result = confirm('Sure you want to delete this task?');
        if (result) {
          onDelete(this.props.task);
        }
      } else {
        onDelete(this.props.task);
      }
      if (onExpand) {
        onExpand(false);
      }
    }
  };

  public handleAddSubtask = (subtask: Subtask) => {
    if (this.props.onEdit) {
      this.props.onEdit({
        ...this.props.task,
        subtasks: [...this.props.task.subtasks, subtask],
      });
    }
  };

  public handleEditSubtask = (subtask: Subtask) => {
    if (this.props.onEdit) {
      this.props.onEdit({
        ...this.props.task,
        subtasks: this.props.task.subtasks.map(
          (st) => (st.id === subtask.id ? subtask : st),
        ),
      });
    }
  };

  public handleRemoveSubtask = (subtask: Subtask) => {
    if (this.props.onEdit) {
      this.props.onEdit({
        ...this.props.task,
        subtasks: this.props.task.subtasks.filter((st) => st.id !== subtask.id),
      });
    }
  };

  public handleChangeMode = () => {
    this.setState({
      mode: this.state.mode === Mode.default ? Mode.editing : Mode.default,
    });
  };

  public handleSelectDeadline = (date: Date) => {
    this.setState({ showCalendar: false }, () => {
      if (this.props.onEdit) {
        this.props.onEdit(setDeadline(this.props.task, date));
      }
    });
  };

  public render() {
    const { task, expand } = this.props;
    return (
      <Container active={expand} completed={task.completed}>
        <header className={styles.taskHeader}>
          <input
            type="checkbox"
            className={styles.checkbox}
            checked={task.completed}
            onChange={this.handleToggle}
            data-testid="checkbox"
          />
          <Title
            active={expand}
            completed={task.completed}
            className={styles.taskTitle}
            value={task.title || ''}
            placeholder="Title..."
            onClick={this.handleExpand}
            onChange={this.handleChangeTitle}
            autoFocus={expand}
            data-testid="title"
          />
          {!expand && !isEmpty(task.tags) ? (
            <Icon glyph={tag} size={18} className={styles.tagIcon} />
          ) : null}
          <Icon
            glyph={expand ? arrow_up : arrow_down}
            style={{ marginLeft: 10 }}
            className={styles.detailsIcon}
            onClick={this.handleToggleDetails}
            data-testid="expand"
          />
        </header>
        <AnimatedDetails
          pose={expand ? 'open' : 'closed'}
          style={{ overflow: 'hidden' }}
        >
          <Details>
            {task.completed ? (
              <pre className={styles.taskDescriptionCompleted}>
                {task.description || 'There are no notes.'}
              </pre>
            ) : (
              <TextArea
                className={styles.taskDescription}
                placeholder="Notes..."
                value={task.description || ''}
                onChange={this.handleChangeDescription}
                data-testid="description"
              />
            )}
            <SubtasksEditor
              subtasks={task.subtasks}
              onCreate={this.handleAddSubtask}
              onEdit={this.handleEditSubtask}
              onRemove={this.handleRemoveSubtask}
              focused={expand}
              isEmpty={
                this.state.mode === Mode.editing && isEmpty(task.subtasks)
              }
            />
            <footer className={styles.taskFooter}>
              {this.props.task.deadline ? (
                <DateTime date={this.props.task.deadline} />
              ) : null}
              <TagsEditor
                className={styles.footerTags}
                tags={task.tags as Tag[]}
                onAddTag={this.handleAddTag}
                onRemoveTags={this.handleRemoveTags}
              />
              {this.props.onDelete && (
                <Icon
                  glyph={trashbin}
                  size={16}
                  className={styles.footerIconDangerous}
                  onClick={this.handleDelete}
                  data-testid="delete"
                />
              )}
              {isEmpty(task.subtasks) && (
                <Icon
                  glyph={list}
                  size={16}
                  className={styles.footerIcon}
                  onClick={this.handleChangeMode}
                />
              )}
              <Icon
                glyph={flag}
                size={16}
                className={styles.footerIcon}
                onClick={this.handleToggleCalendar}
              />
              {this.state.showCalendar && (
                <Calendar onSelectDate={this.handleSelectDeadline} />
              )}
            </footer>
          </Details>
        </AnimatedDetails>
      </Container>
    );
  }
}

const Container = styled<{ active?: boolean; completed?: boolean }, 'div'>(
  'div',
)`
  overflow: hidden;
  box-sizing: border-box;
  position: relative;
  padding: 15px 10px;
  background-color: ${({ active }) => (active ? '#fff' : 'transparent')};
  color: ${({ completed }) => (completed ? 'lightgrey' : '#334')};
  box-shadow: ${({ active }) =>
    active ? '0 1px 3px 2px rgba(0,0,0,0.1)' : 'none'};
`;

const Title = styled(TextArea)`
  text-decoration: ${({ completed }: any) =>
    completed ? 'line-through' : 'none'};
  color: ${({ active }: any) => (active ? '#112' : '#445')};
  color: ${({ completed }: any) => (completed ? '#b3b3bb' : null)};
`;

const AnimatedDetails = posed.div({
  open: {
    height: 'auto',
    opacity: 1,
    transition: {
      duration: 200,
    },
  },
  closed: {
    height: 0,
    opacity: 0,
    transition: {
      duration: 200,
    },
  },
});

const Details = styled.div`
  overflow: hidden;
  box-sizing: border-box;
  margin-left: calc(1em + 20px);
  padding: 10px 0 0;
`;

export default Task;
