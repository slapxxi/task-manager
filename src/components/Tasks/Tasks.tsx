import { Task as ITask } from '@local/types';
import { isEmpty } from 'lodash';
import React from 'react';
import styles from './styles.css';

interface Props {
  tasks: ITask[];
  children?: (params: { task: ITask; index: number }) => React.ReactNode;
}

function Tasks({ tasks, children }: Props) {
  if (isEmpty(tasks)) {
    return <p>There are no tasks yet.</p>;
  }
  return (
    <ul className={styles.tasks}>
      {tasks.map((t, index) => (
        <li key={t.id}>{children && children({ task: t, index })}</li>
      ))}
    </ul>
  );
}

export default React.memo(Tasks);
