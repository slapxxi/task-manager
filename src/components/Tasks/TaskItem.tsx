import * as React from 'react';
import posed from 'react-pose';
import styled from 'styled-components';
import { Icon } from '../';
import dots from '../../assets/dots.svg';
import flag from '../../assets/flag.svg';
import list from '../../assets/list.svg';
import trashbin from '../../assets/trashbin.svg';
import { toggleTask } from '../../lib/tasks';
import TextArea from '../TextArea/TextArea';
import styles from './styles.css';

interface Props {
  task: Task;
  expand?: boolean;
  confirmDelete?: boolean;
  onChange?: (task: Task) => void;
  onDelete?: (task: Task) => void;
  onExpand?: (expand: boolean) => void;
}

class TaskItem extends React.PureComponent<Props, {}> {
  constructor(props: Props) {
    super(props);
    this.state = { showDetails: this.props.expand ? true : false };
  }

  public handleToggle = () => {
    const { task, onChange } = this.props;
    if (onChange) {
      onChange(toggleTask(task));
    }
  };

  public handleChangeDescription = (value: string) => {
    const { task, onChange } = this.props;
    if (onChange) {
      onChange({ ...task, description: value });
    }
  };

  public handleChangeTitle = (title: string) => {
    const { task, onChange } = this.props;
    if (onChange) {
      onChange({ ...task, title });
    }
  };

  public handleToggleDetails = () => {
    if (this.props.onExpand) {
      this.props.onExpand(!this.props.expand);
    }
  };

  public handleDelete = () => {
    const { onDelete, confirmDelete } = this.props;
    if (onDelete) {
      if (confirmDelete) {
        const result = confirm('Sure you want to delete this task?');
        if (result) {
          onDelete(this.props.task);
        }
      } else {
        onDelete(this.props.task);
      }
    }
  };

  public render() {
    const { task, expand } = this.props;
    return (
      // @ts-ignore
      <Container active={expand} completed={task.completed}>
        <header className={styles.taskHeader}>
          <input
            type="checkbox"
            className={styles.checkbox}
            checked={task.completed}
            onChange={this.handleToggle}
            data-testid="checkbox"
          />
          <Title
            // @ts-ignore
            completed={task.completed}
            active={expand}
            className={styles.taskTitle}
            type="text"
            value={task.title || ''}
            placeholder="Title..."
            onChange={this.handleChangeTitle}
            data-testid="title"
          />
          {expand ? null : <Tags />}
          <Icon
            glyph={dots}
            width={15}
            style={{ marginLeft: 10 }}
            className={styles.detailsIcon}
            onClick={this.handleToggleDetails}
          />
        </header>
        <AnimatedDetails
          pose={expand ? 'open' : 'closed'}
          style={{ overflow: 'hidden' }}
        >
          <Details>
            {task.completed ? (
              <pre className={styles.taskDescriptionCompleted}>
                {task.description || 'There are no notes.'}
              </pre>
            ) : (
              <TextArea
                className={styles.taskDescription}
                placeholder="Notes..."
                value={task.description || ''}
                onChange={this.handleChangeDescription}
                data-testid="description"
              />
            )}
            <footer className={styles.taskFooter}>
              <Tags className={styles.footerTags} />
              {this.props.onDelete && (
                <Icon
                  glyph={trashbin}
                  size={16}
                  className={styles.footerIconDangerous}
                  onClick={this.handleDelete}
                  data-testid="delete"
                />
              )}
              <Icon glyph={list} size={16} className={styles.footerIcon} />
              <Icon glyph={flag} size={16} className={styles.footerIcon} />
            </footer>
          </Details>
        </AnimatedDetails>
      </Container>
    );
  }
}

function Tags(props: any) {
  return (
    <ul className={styles.tags} {...props}>
      <li className={styles.tag}>html</li>
      <li className={styles.tag}>css</li>
    </ul>
  );
}

const Container = styled<{ active: boolean; completed: boolean }, 'li'>('li')`
  overflow: hidden;
  box-sizing: border-box;
  position: relative;
  padding: 15px 10px;
  background-color: ${({ active }) => (active ? '#fff' : 'transparent')};
  color: ${({ completed }) => (completed ? 'lightgrey' : '#334')};
  box-shadow: ${({ active }) =>
    active ? '0 1px 3px 2px rgba(0,0,0,0.1)' : 'none'};
`;

const Title = styled(TextArea)`
  text-decoration: ${({ completed }: any) =>
    completed ? 'line-through' : 'none'};
  color: ${({ active }: any) => (active ? '#112' : '#445')};
  color: ${({ completed }: any) => (completed ? '#b3b3bb' : null)};
`;

const AnimatedDetails = posed.div({
  open: {
    height: 'auto',
    opacity: 1,
    transition: {
      duration: 200,
    },
  },
  closed: {
    height: 0,
    opacity: 0,
    transition: {
      duration: 200,
    },
  },
});

const Details = styled.div`
  overflow: hidden;
  box-sizing: border-box;
  margin-left: calc(1em + 20px);
  padding: 10px 0 0;
`;

export default TaskItem;
