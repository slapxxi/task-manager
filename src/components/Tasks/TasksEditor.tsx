import { createTask } from '@lib';
import { Action, Task as ITask } from '@local/types';
import React, { useReducer } from 'react';
import { Button } from '../';
import { isValidTask } from '../../lib/tasks';
import Task from './Task';
import Tasks from './Tasks';

enum Mode {
  default,
  create,
}

enum Actions {
  create,
  finish,
  expand,
  hide,
  cancel,
}

interface Props {
  tasks: ITask[];
  onCreate?: (task: ITask) => void;
  onEdit?: (task: ITask) => void;
  onDelete?: (task: ITask) => void;
}

interface State {
  mode: Mode;
  activeIndex: number;
  lastActiveIndex: number;
}

function TasksEditor({ tasks, onEdit, onDelete, onCreate }: Props) {
  const [state, dispatch] = useReducer(tasksReducer, {
    mode: Mode.default,
    activeIndex: NaN,
    lastActiveIndex: NaN,
  });

  function handleCreateTask(task: ITask) {
    if (isValidTask(task)) {
      dispatch({ type: Actions.finish, payload: tasks.length });
      if (onCreate) {
        onCreate(task);
      }
    }
  }

  function handleEdit(task: ITask) {
    if (onEdit) {
      onEdit(task);
    }
  }

  function handleDelete(task: ITask) {
    if (onDelete) {
      onDelete(task);
    }
  }

  function handleEnableCreate() {
    dispatch({ type: Actions.create });
  }

  function handleCancel() {
    dispatch({ type: Actions.cancel });
  }

  function handleExpand(index: number, expand: boolean) {
    if (expand === false) {
      dispatch({ type: Actions.hide });
      return;
    }
    dispatch({ type: Actions.expand, payload: index });
  }

  return (
    <>
      <Tasks tasks={state.mode === Mode.create ? [...tasks, createTask({})] : tasks}>
        {({ task, index }) =>
          state.mode === Mode.create && index === tasks.length ? (
            <Task task={task} expand={true} onEdit={handleCreateTask} />
          ) : (
            <Task
              task={task}
              expand={index === state.activeIndex}
              onEdit={handleEdit}
              onDelete={handleDelete}
              onExpand={(expand) => handleExpand(index, expand)}
              confirmDelete
            />
          )
        }
      </Tasks>
      {state.mode === Mode.default ? (
        <Button onClick={handleEnableCreate}>Create Task</Button>
      ) : (
        <Button onClick={handleCancel}>Cancel</Button>
      )}
    </>
  );
}

function tasksReducer(state: State, action: Action<Actions>) {
  switch (action.type) {
    case Actions.create:
      return {
        ...state,
        mode: Mode.create,
        lastActiveIndex: state.activeIndex,
        activeIndex: NaN,
      };
    case Actions.finish:
      return { ...state, mode: Mode.default, activeIndex: action.payload };
    case Actions.expand:
      return { ...state, mode: Mode.default, activeIndex: action.payload };
    case Actions.hide:
      return { ...state, activeIndex: NaN };
    case Actions.cancel:
      return {
        ...state,
        mode: Mode.default,
        activeIndex: state.lastActiveIndex,
        lastActiveIndex: NaN,
      };
    default:
      return state;
  }
}

export default React.memo(TasksEditor);
