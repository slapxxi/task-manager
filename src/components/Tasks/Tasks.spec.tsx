import React from 'react';
import { render } from 'react-testing-library';
import { createTask } from '../../lib/tasks';
import Tasks from './Tasks';

const tasks = [createTask({ title: 'first' }), createTask({ title: 'second' })];

it('renders tasks', () => {
  const { container } = render(
    <Tasks tasks={tasks}>{({ task }) => <MockTask task={task} />}</Tasks>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders information when tasks are empty', () => {
  const { container } = render(<Tasks tasks={[]} />);
  expect(container.firstChild).toMatchSnapshot();
});

function MockTask({ task }: any) {
  return <div>{task.title}</div>;
}
