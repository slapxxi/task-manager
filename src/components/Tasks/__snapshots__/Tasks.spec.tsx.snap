// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders information when tasks are empty 1`] = `
<p>
  There are no tasks yet.
</p>
`;

exports[`renders tasks 1`] = `
<ul
  class="tasks"
>
  <li>
    <div>
      first
    </div>
  </li>
  <li>
    <div>
      second
    </div>
  </li>
</ul>
`;
